In [153]:
from google.colab import drive
drive.mount('/content/gdrive')

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [154]:
import os

import numpy as np

from matplotlib import pyplot as plt


import torch
from torch import optim
from torch import nn
from torch.utils.data import DataLoader

import torchvision
import torchvision.transforms as transforms
import torch.nn.functional as F
import torchvision.datasets as datasets
from tqdm import tqdm

from sklearn.model_selection import train_test_split
from torch.utils.data import DataLoader, TensorDataset, Dataset

%matplotlib inline

In [155]:
class RedNeuronalMLS(nn.Module):
    def __init__(self, input_size, num_classes):
        super(RedNeuronalMLS, self).__init__()

        self.fc1 = nn.Linear(input_size, 5000)
        self.fc2 = nn.Linear(5000, num_classes)

    def forward(self, x):

        x = self.fc1(x)
        x = torch.sigmoid(x)

        x = self.fc2(x)
        return x

In [156]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(device)

input_size = 112
num_classes = 5
learning_rate = 0.001
batch_size = 1000
num_epochs = 7

cpu


In [157]:
class CustomDataset(Dataset):
    def __init__(self, file_path):

        data = np.loadtxt(file_path, delimiter=',', skiprows=1)
        self.X = torch.tensor(data[:, 0:], dtype=torch.float32)
        self.y = torch.tensor(data[:, 111], dtype=torch.long)

    def __len__(self):
        return len(self.X)

    def __getitem__(self, idx):
        return self.X[idx], self.y[idx]


file_path = "/content/gdrive/MyDrive/Colab Notebooks/Datasets/l4d2_player_stats_preparado.csv"


dataset = CustomDataset(file_path)


train_size = int(0.8 * len(dataset))
test_size = len(dataset) - train_size
train_dataset, test_dataset = torch.utils.data.random_split(dataset, [train_size, test_size])


batch_size = 1000
train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False)

In [158]:
print(len(train_dataset))
print(len(test_dataset))

16664
4166


In [159]:
model = RedNeuronalMLS(input_size=input_size, num_classes=num_classes).to(device)

In [160]:
model

RedNeuronalMLS(
  (fc1): Linear(in_features=112, out_features=5000, bias=True)
  (fc2): Linear(in_features=5000, out_features=5, bias=True)
)

In [161]:
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=learning_rate)

In [162]:
for epoch in range(num_epochs):
    for batch_idx, (data, targets) in enumerate(tqdm(train_loader)):

        data = data.to(device=device)
        targets = targets.to(device=device)

        data = data.reshape(data.shape[0], -1)
        print(data.shape)

        scores = model(data)
        loss = criterion(scores, targets)

        optimizer.zero_grad()
        loss.backward()

        optimizer.step()

  6%|▌         | 1/17 [00:00<00:02,  7.20it/s]

torch.Size([1000, 112])
torch.Size([1000, 112])


 12%|█▏        | 2/17 [00:00<00:02,  7.11it/s]

torch.Size([1000, 112])


 18%|█▊        | 3/17 [00:00<00:01,  7.08it/s]

torch.Size([1000, 112])


 24%|██▎       | 4/17 [00:00<00:01,  7.08it/s]

torch.Size([1000, 112])


 29%|██▉       | 5/17 [00:00<00:01,  6.66it/s]

torch.Size([1000, 112])


 35%|███▌      | 6/17 [00:00<00:01,  6.85it/s]

torch.Size([1000, 112])


 41%|████      | 7/17 [00:01<00:01,  6.75it/s]

torch.Size([1000, 112])


 47%|████▋     | 8/17 [00:01<00:01,  6.57it/s]

torch.Size([1000, 112])


 53%|█████▎    | 9/17 [00:01<00:01,  6.75it/s]

torch.Size([1000, 112])


 59%|█████▉    | 10/17 [00:01<00:01,  6.90it/s]

torch.Size([1000, 112])


 65%|██████▍   | 11/17 [00:01<00:00,  7.06it/s]

torch.Size([1000, 112])


 71%|███████   | 12/17 [00:01<00:00,  7.18it/s]

torch.Size([1000, 112])


 76%|███████▋  | 13/17 [00:01<00:00,  6.94it/s]

torch.Size([1000, 112])


 82%|████████▏ | 14/17 [00:02<00:00,  7.07it/s]

torch.Size([1000, 112])


 88%|████████▊ | 15/17 [00:02<00:00,  7.16it/s]

torch.Size([1000, 112])


 94%|█████████▍| 16/17 [00:02<00:00,  7.08it/s]

torch.Size([664, 112])


  0%|          | 0/17 [00:00<?, ?it/s]

torch.Size([1000, 112])


  6%|▌         | 1/17 [00:00<00:02,  5.39it/s]

torch.Size([1000, 112])


 12%|█▏        | 2/17 [00:00<00:02,  5.53it/s]

torch.Size([1000, 112])


 18%|█▊        | 3/17 [00:00<00:02,  5.50it/s]

torch.Size([1000, 112])


 24%|██▎       | 4/17 [00:00<00:02,  5.49it/s]

torch.Size([1000, 112])


 29%|██▉       | 5/17 [00:00<00:02,  5.46it/s]

torch.Size([1000, 112])


 41%|████      | 7/17 [00:01<00:01,  5.26it/s]

torch.Size([1000, 112])
torch.Size([1000, 112])


 53%|█████▎    | 9/17 [00:01<00:01,  5.40it/s]

torch.Size([1000, 112])
torch.Size([1000, 112])


 65%|██████▍   | 11/17 [00:02<00:01,  5.32it/s]

torch.Size([1000, 112])
torch.Size([1000, 112])


 76%|███████▋  | 13/17 [00:02<00:00,  5.50it/s]

torch.Size([1000, 112])
torch.Size([1000, 112])


 88%|████████▊ | 15/17 [00:02<00:00,  5.41it/s]

torch.Size([1000, 112])
torch.Size([1000, 112])


100%|██████████| 17/17 [00:03<00:00,  5.46it/s]


torch.Size([664, 112])


  0%|          | 0/17 [00:00<?, ?it/s]

torch.Size([1000, 112])


  6%|▌         | 1/17 [00:00<00:02,  5.54it/s]

torch.Size([1000, 112])


 12%|█▏        | 2/17 [00:00<00:02,  5.41it/s]

torch.Size([1000, 112])


 24%|██▎       | 4/17 [00:00<00:02,  5.19it/s]

torch.Size([1000, 112])
torch.Size([1000, 112])


 35%|███▌      | 6/17 [00:01<00:01,  6.06it/s]

torch.Size([1000, 112])
torch.Size([1000, 112])


 47%|████▋     | 8/17 [00:01<00:01,  6.61it/s]

torch.Size([1000, 112])
torch.Size([1000, 112])


 59%|█████▉    | 10/17 [00:01<00:01,  6.73it/s]

torch.Size([1000, 112])
torch.Size([1000, 112])


 71%|███████   | 12/17 [00:01<00:00,  6.73it/s]

torch.Size([1000, 112])
torch.Size([1000, 112])


 82%|████████▏ | 14/17 [00:02<00:00,  6.89it/s]

torch.Size([1000, 112])
torch.Size([1000, 112])


 94%|█████████▍| 16/17 [00:02<00:00,  7.14it/s]

torch.Size([1000, 112])
torch.Size([664, 112])


  6%|▌         | 1/17 [00:00<00:02,  7.65it/s]

torch.Size([1000, 112])
torch.Size([1000, 112])


 18%|█▊        | 3/17 [00:00<00:01,  7.41it/s]

torch.Size([1000, 112])
torch.Size([1000, 112])


 29%|██▉       | 5/17 [00:00<00:01,  7.16it/s]

torch.Size([1000, 112])
torch.Size([1000, 112])


 41%|████      | 7/17 [00:00<00:01,  7.17it/s]

torch.Size([1000, 112])
torch.Size([1000, 112])


 53%|█████▎    | 9/17 [00:01<00:01,  7.13it/s]

torch.Size([1000, 112])
torch.Size([1000, 112])


 65%|██████▍   | 11/17 [00:01<00:00,  7.05it/s]

torch.Size([1000, 112])
torch.Size([1000, 112])


 76%|███████▋  | 13/17 [00:01<00:00,  6.21it/s]

torch.Size([1000, 112])
torch.Size([1000, 112])


 88%|████████▊ | 15/17 [00:02<00:00,  4.22it/s]

torch.Size([1000, 112])
torch.Size([1000, 112])


100%|██████████| 17/17 [00:02<00:00,  6.15it/s]


torch.Size([664, 112])


  0%|          | 0/17 [00:00<?, ?it/s]

torch.Size([1000, 112])


  6%|▌         | 1/17 [00:00<00:03,  5.20it/s]

torch.Size([1000, 112])


 12%|█▏        | 2/17 [00:00<00:02,  5.37it/s]

torch.Size([1000, 112])


 18%|█▊        | 3/17 [00:00<00:02,  5.47it/s]

torch.Size([1000, 112])


 24%|██▎       | 4/17 [00:00<00:02,  5.53it/s]

torch.Size([1000, 112])


 29%|██▉       | 5/17 [00:00<00:02,  5.61it/s]

torch.Size([1000, 112])


 35%|███▌      | 6/17 [00:01<00:01,  5.75it/s]

torch.Size([1000, 112])


 41%|████      | 7/17 [00:01<00:01,  5.81it/s]

torch.Size([1000, 112])


 47%|████▋     | 8/17 [00:01<00:01,  5.69it/s]

torch.Size([1000, 112])


 53%|█████▎    | 9/17 [00:01<00:01,  5.63it/s]

torch.Size([1000, 112])


 59%|█████▉    | 10/17 [00:01<00:01,  5.41it/s]

torch.Size([1000, 112])

 65%|██████▍   | 11/17 [00:01<00:01,  5.41it/s]


torch.Size([1000, 112])


 76%|███████▋  | 13/17 [00:02<00:00,  5.56it/s]

torch.Size([1000, 112])
torch.Size([1000, 112])


 88%|████████▊ | 15/17 [00:02<00:00,  5.88it/s]

torch.Size([1000, 112])
torch.Size([1000, 112])


100%|██████████| 17/17 [00:02<00:00,  5.87it/s]


torch.Size([664, 112])


  0%|          | 0/17 [00:00<?, ?it/s]

torch.Size([1000, 112])


  6%|▌         | 1/17 [00:00<00:02,  7.77it/s]

torch.Size([1000, 112])


 12%|█▏        | 2/17 [00:00<00:01,  7.63it/s]

torch.Size([1000, 112])


 18%|█▊        | 3/17 [00:00<00:01,  7.41it/s]

torch.Size([1000, 112])


 24%|██▎       | 4/17 [00:00<00:01,  6.77it/s]

torch.Size([1000, 112])


 29%|██▉       | 5/17 [00:00<00:01,  6.92it/s]

torch.Size([1000, 112])


 35%|███▌      | 6/17 [00:00<00:01,  7.21it/s]

torch.Size([1000, 112])


 41%|████      | 7/17 [00:00<00:01,  7.28it/s]

torch.Size([1000, 112])


 47%|████▋     | 8/17 [00:01<00:01,  7.25it/s]

torch.Size([1000, 112])


 53%|█████▎    | 9/17 [00:01<00:01,  7.42it/s]

torch.Size([1000, 112])


 59%|█████▉    | 10/17 [00:01<00:00,  7.37it/s]

torch.Size([1000, 112])


 65%|██████▍   | 11/17 [00:01<00:00,  7.40it/s]

torch.Size([1000, 112])


 71%|███████   | 12/17 [00:01<00:00,  7.41it/s]

torch.Size([1000, 112])


 76%|███████▋  | 13/17 [00:01<00:00,  7.48it/s]

torch.Size([1000, 112])


 82%|████████▏ | 14/17 [00:01<00:00,  7.57it/s]

torch.Size([1000, 112])


 88%|████████▊ | 15/17 [00:02<00:00,  7.62it/s]

torch.Size([1000, 112])


 94%|█████████▍| 16/17 [00:02<00:00,  7.56it/s]

torch.Size([664, 112])


  0%|          | 0/17 [00:00<?, ?it/s]

torch.Size([1000, 112])


  6%|▌         | 1/17 [00:00<00:02,  7.59it/s]

torch.Size([1000, 112])


 12%|█▏        | 2/17 [00:00<00:02,  7.20it/s]

torch.Size([1000, 112])


 18%|█▊        | 3/17 [00:00<00:02,  6.98it/s]

torch.Size([1000, 112])


 24%|██▎       | 4/17 [00:00<00:01,  6.76it/s]

torch.Size([1000, 112])


 29%|██▉       | 5/17 [00:00<00:01,  7.00it/s]

torch.Size([1000, 112])


 35%|███▌      | 6/17 [00:00<00:01,  7.21it/s]

torch.Size([1000, 112])


 41%|████      | 7/17 [00:00<00:01,  7.36it/s]

torch.Size([1000, 112])


 47%|████▋     | 8/17 [00:01<00:01,  7.44it/s]

torch.Size([1000, 112])


 53%|█████▎    | 9/17 [00:01<00:01,  7.53it/s]

torch.Size([1000, 112])


 65%|██████▍   | 11/17 [00:01<00:01,  5.49it/s]

torch.Size([1000, 112])
torch.Size([1000, 112])


 76%|███████▋  | 13/17 [00:01<00:00,  6.43it/s]

torch.Size([1000, 112])
torch.Size([1000, 112])


 82%|████████▏ | 14/17 [00:02<00:00,  6.73it/s]

torch.Size([1000, 112])


 94%|█████████▍| 16/17 [00:02<00:00,  5.33it/s]

torch.Size([1000, 112])
torch.Size([664, 112])


100%|██████████| 17/17 [00:02<00:00,  6.43it/s]


In [163]:
def check_accuracy(loader, model):
    num_correct = 0
    num_samples = 0
    model.eval()

    predicciones = []
    with torch.no_grad():
        for x, y in loader:
            x = x.to(device=device)
            y = y.to(device=device)
            x = x.reshape(x.shape[0], -1)

            scores = model(x)
            _, predictions = scores.max(1)
            predicciones.append(predictions)

            num_correct += (predictions == y).sum()
            num_samples += predictions.size(0)

    model.train()
    return num_correct/num_samples, predicciones

p_train, pred_train  = check_accuracy(train_loader, model)
p_test, pred_test  = check_accuracy(test_loader, model)

print(f"Accuracy on training set: {p_train*100:.2f}")
print(f"Accuracy on test set: {p_test*100:.2f}")

Accuracy on training set: 53.65
Accuracy on test set: 53.67


son los mejores resultados obtenidos, 53.65 para el training set y 53.67 para el test set, usando:
input_size = 112
num_classes = 5
learning_rate = 0.001
batch_size = 1000
num_epochs = 7
y 5000 nodos de capa oculta